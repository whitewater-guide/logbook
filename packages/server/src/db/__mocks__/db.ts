import {
  DatabasePoolConnectionType,
  DatabasePoolType,
  createPool,
} from 'slonik';

import { Client } from 'pg';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: `logbook_test_${process.env.JEST_WORKER_ID}`,
  password: process.env.POSTGRES_PASSWORD,
};

let _pool: DatabasePoolType | undefined;

export const db: () => DatabasePoolConnectionType = () => {
  if (!_pool) {
    throw new Error('no test pool found');
  }
  return _pool;
};

export const setupDB = async () => {
  const workerID = process.env.JEST_WORKER_ID;
  if (!workerID) {
    throw new Error('JEST_WORKER_ID not found');
  }
  const client = new Client({
    host: process.env.POSTGRES_HOST,
    user: process.env.POSTGRES_USER,
    database: process.env.POSTGRES_DB,
    password: process.env.POSTGRES_PASSWORD,
  });
  try {
    await client.connect();
    await client.query(`DROP DATABASE IF EXISTS logbook_test_${workerID}`);
    await client.query(
      `CREATE DATABASE logbook_test_${workerID} TEMPLATE logbook_test_template`,
    );
  } finally {
    await client.end();
  }

  _pool = createPool(
    `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  );
};

export const teardownDB = async () => {
  await _pool?.end();
};
