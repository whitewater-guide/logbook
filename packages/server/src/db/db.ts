import { createPool, DatabasePoolConnectionType } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
