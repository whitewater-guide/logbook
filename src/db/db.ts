import {
  DatabasePoolConnectionType,
  createPool,
  createTypeParserPreset,
} from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
