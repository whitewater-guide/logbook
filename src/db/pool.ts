import { createPool, createTypeParserPreset } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database:
    process.env.NODE_ENV === 'test'
      ? `logbook_test_${process.env.JEST_WORKER_ID}`
      : process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

export const db = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);
