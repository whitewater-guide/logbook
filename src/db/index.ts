import { db } from './pool';

export default db;
