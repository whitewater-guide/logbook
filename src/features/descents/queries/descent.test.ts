import { DESCENT_1, DESCENT_2, USER_1, USER_2 } from '~/test/fixtures';
import {
  GetDescentQuery,
  GetDescentQueryVariables,
} from './descent.test.generated';
import { setupDB, teardownDB } from '~/db';

import { DESCENT_2_SHARE_TOKEN } from '../../../test/fixtures';
import { gql } from 'apollo-server';
import { runQuery } from '~/test/apollo-helpers';

beforeEach(setupDB);
afterEach(teardownDB);

const query = gql`
  query getDescent($id: ID, $shareToken: String) {
    descent(id: $id, shareToken: $shareToken) {
      id
      userId

      section {
        id

        region
        river
        section
        difficulty

        upstreamId
        upstreamData

        createdAt
        updatedAt
      }

      startedAt
      duration
      level {
        value
        unit
      }
      comment
      public

      upstreamData

      createdAt
      updatedAt
    }
  }
`;

describe('permissions', () => {
  type PermissionsTestCase = [
    string,
    GetDescentQueryVariables,
    string | undefined,
    boolean,
  ];

  it.each<PermissionsTestCase>([
    ['anon should get public descent', { id: DESCENT_1 }, undefined, true],
    [
      'anon should not get private descent',
      { id: DESCENT_2 },
      undefined,
      false,
    ],
    ['user should get public descent', { id: DESCENT_1 }, USER_1, true],
    ['user should get own private descent', { id: DESCENT_2 }, USER_1, true],
    [
      'user should not get other users private descent',
      { id: DESCENT_2 },
      USER_2,
      false,
    ],
    [
      'user should get other users private descent with share token',
      { shareToken: DESCENT_2_SHARE_TOKEN },
      USER_2,
      true,
    ],
    [
      'anon should not get other users private descent with share token',
      { shareToken: DESCENT_2_SHARE_TOKEN },
      undefined,
      false,
    ],
  ])('%s', async (_, vars, uid, allowed) => {
    const result = await runQuery<GetDescentQuery, GetDescentQueryVariables>(
      query,
      vars,
      uid,
    );
    if (allowed) {
      expect(result.errors).toBeUndefined();
      expect(result.data?.descent).toBeTruthy();
    } else {
      expect(result.errors).toBeTruthy();
      expect(result.data?.descent).toBeNull();
    }
  });
});

it('should return descent', async () => {
  const result = await runQuery(query, { id: DESCENT_1 });
  expect(result.errors).toBeUndefined();
  expect(result.data!.descent).toMatchSnapshot();
});
