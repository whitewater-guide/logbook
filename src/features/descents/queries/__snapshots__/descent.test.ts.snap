// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return descent 1`] = `
Object {
  "comment": "descent 1 comment",
  "createdAt": 2020-01-01T00:00:00.000Z,
  "duration": 3600,
  "id": "97998f67-74f2-43db-a4e4-198c2ca540a7",
  "level": Object {
    "unit": "m3/s",
    "value": 99.9,
  },
  "public": true,
  "section": Object {
    "createdAt": 2020-01-01T00:00:00.000Z,
    "difficulty": 7,
    "id": "45e70501-5d20-449f-b854-52b06b138831",
    "region": "Norway",
    "river": "Sjoa",
    "section": "Amot",
    "updatedAt": 2020-01-01T00:00:00.000Z,
    "upstreamData": Object {
      "upstream": "data",
    },
    "upstreamId": "__upstream__id__",
  },
  "startedAt": 2020-01-01T00:00:00.000Z,
  "updatedAt": 2020-01-01T00:00:00.000Z,
  "upstreamData": Object {
    "upstream": "data",
  },
  "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
}
`;
